File: university-service/src/main/java/ua/com/foxminded/courseproject/controllers/PersonController.java
package ua.com.foxminded.courseproject.controllers;

import org.springframework.data.domain.Pageable;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import ua.com.foxminded.courseproject.dto.PersonDto;
import ua.com.foxminded.courseproject.service.PersonService;

import java.util.UUID;

public class PersonController<T extends PersonDto, S extends PersonService> {

    S service;

    protected ResponseEntity getPersons(Pageable pageable) {
        return new ResponseEntity(service.findAll(pageable), HttpStatus.OK);
    }

    protected ResponseEntity getPersonById(UUID id) {
        T person = (T) service.findById(id);
        return new ResponseEntity<>(person, HttpStatus.OK);
    }

    protected ResponseEntity deletePersonById(UUID id) {
        service.delete(id);
        return new ResponseEntity<>(HttpStatus.NO_CONTENT);
    }
}

File: university-service/src/main/kotlin/ua/com/foxminded/courseproject/controllers/PersonController.kt
package ua.com.foxminded.courseproject.controllers

import java.util.UUID
import org.springframework.http.HttpStatus
import org.springframework.http.ResponseEntity
import reactor.core.publisher.Mono
import ua.com.foxminded.courseproject.dto.PersonDto
import ua.com.foxminded.courseproject.service.PersonService

open class PersonController<T : PersonDto, S : PersonService<*>> {

    lateinit var service: S
    protected fun getPersons(): ResponseEntity<*> {
        return ResponseEntity<Any?>(service.findAll(), HttpStatus.OK)
    }

    protected fun getPersonById(id: UUID): ResponseEntity<*> {
        val person = service.findById(id) as Mono<T>
        return ResponseEntity(person, HttpStatus.OK)
    }

    protected fun deletePersonById(id: UUID): Mono<ResponseEntity<*>> {
        return service.delete(id).thenReturn(ResponseEntity.status(HttpStatus.NO_CONTENT).build<Void>())
    }
}
